File: basic/merge_sort.js
function mergeSort(arr) {
  if (arr.length>1) {
    let mid = Math.floor(arr.length/2)
    let leftArr = arr.slice(0, mid)
    let rightArr = arr.slice(mid)
    mergeSort(leftArr)
    mergeSort(rightArr)
    let i = j = k = 0
    while(i < leftArr.length && j < rightArr.length) {
      if(leftArr[i] < rightArr[j]) {
        // arr[i] = leftArr[i]
        arr[k] = leftArr[i]
        k++
        i++
      } else {
        arr[k] = rightArr[j]
        k++
        j++
      }
    }
    while (j < rightArr.length) {
      arr[k] = rightArr[j]
      j++
      k++
    }
    while (i < leftArr.length) {
      arr[k] = leftArr[i]
      i++
      k++
    }
  }
}

let arr = [38, 43, 27, 3, 9, 82, 10]
mergeSort(arr)
console.log(arr)

File: basic/sorting/merge_sort.py
def merge_sort(arr):
  if len(arr) > 1:
    mid = len(arr)//2
    L = arr[:mid]
    R = arr[mid:]
    merge_sort(L)
    merge_sort(R)

    i = j = k = 0
    while i < len(L) and j < len(R):
      if L[i] < R[j]:
        arr[k] = L[i]
        i += 1
      else:
        arr[k] = R[j]
        j += 1
      k += 1
      print(arr)
    
    while i < len(L):
      arr[k] = L[i]
      k += 1
      i += 1
    while j < len(R):
      arr[k] = R[j]
      k += 1
      j += 1

arr = [38, 27, 82, 3, 9, 43, 10]
merge_sort(arr)
print(arr)
